### ISSUE TITLE
Should clients really get notifications if other clients reply?

### ISSUE THREAD

**Issue by fiedl** [ivaldi/brimir#259]
I've just done a `git pull upstream master` and discovered some maybe disturbing change, I would like to discuss. Maybe this change is unintended, or, I can't see the obvious solution.

## Scenario

* Suppose, **client1** sends his problem to the ticket system via support@example.com.
* **Agent** answers and suggests a possible solution. For some reason, the **agent** adds **client2** (who might be a team member or the boss of *client1*) as recipient for this reply. (`reply_of_agent`)
* **Client1** replies: "Sorry, your solution does not work for me." (`reply_of_client1`)
* **Client2** replies: "Client1 is stupid to ask that question. Did he even switch on his computer?" (`reply_of_client2`)

## Expected Behaviour

Since **Client2**, when he replies, only sees support@example.com in the `To:` field, not **Client1**, he would not expect that **Client2** receives his reply.

`reply_of_client2.notified_users` should not include **Client1**.

## Observed Behaviour

In fact, when **Client2** replies, **Agent** and **Client1** receive notifications.

`reply_of_client2.notified_users` includes **Client1**.

## Cause

I think the cause of this behaviour is here: https://github.com/ivaldi/brimir/blob/c7799f47d07aeb0f31bd5758a8c16d912536c51b/app/models/reply.rb#L52L53

```ruby
# app/model/reply.rb

def set_default_notifications!
    unless reply_to_type.nil?
      self.notified_users =
          ([reply_to.user] + reply_to.notified_users - [user]).uniq
# ...
```

Since for `reply_of_client2`, who replies to the answer of *Agent*, `reply_to` is `reply_of_agent`, `reply_to.notified_users` includes *Client1* and *Client2*. Thus, when *Client2* replies to `reply_of_agent`, *Agent* and *Client1* receive a notification.

I think this has been added in https://github.com/ivaldi/brimir/commit/485a52d5797d562387a9fcee0b67355a8db50bf5.

## Is this intended?

The first question is: Is this behaviour really intended or did this happen by accident?

I would imagine this behaviour could be helpful in some situations where several clients discuss some issue with an agent. But I think it should be transparent to **Client2** who will receive his reply when he replies to support@example.com.

## Discussing the `To:` field

If **Client2** replies `To: support@example.com, client1@example.com`, he knows that **client1** receives this reply. He would be able to remove **client1** from the recipient list.

The **solution that comes to mind** would be to modify the headers of `reply_of_agent` such that **Client1** and **Client2** both are included in the `To:` field. Then, **Client2** could click either on **Reply** or **Reply to all**.

But: From the brimir UI, an agent can't choose between `To` or `BCC` notifications, right?

Also, there might be other implications of having several recipients in one email at once. At least, this requires some more thoughts.

What do you think?

## Test

In order to make the case more transparent, I've written a short test.

```ruby
# test/models/reply_test.rb
# ...
 test 'should not notify other clients when one of the clients replies' do
    agent = User.create! email: 'agent@example.com', agent: true
    client1 = User.create! email: 'client1@example.com'
    client2 = User.create! email: 'client2@example.com'
    
    ticket = Ticket.create from: 'client1@example.com', content: 'This is my problem'
    reply_of_the_agent = ticket.replies.create! content: 'This might be the solution.', user: agent
    reply_of_the_agent.notified_users << client1
    reply_of_the_agent.notified_users << client2  # This client, for some reasons, needs this reply as CC, but client1 does not know this at this point.
    
    reply_of_client1 = ticket.replies.create! content: 'It did not work.', user: client1
    reply_of_client1.reply_to = reply_of_the_agent
    reply_of_client1.set_default_notifications!
    
    reply_of_client2 = ticket.replies.create! content: 'client1 is stupid! Did he even start his computer?', user: client2
    reply_of_client2.reply_to = reply_of_the_agent
    reply_of_client2.set_default_notifications!
    
    assert reply_of_client2.notified_users.include?(agent)
    assert not(reply_of_client2.notified_users.include?(client1))
  end
```

**Comment by fiedl** [comment]
*Internal note*: We did the same thing in https://github.com/fiedl/brimir/blob/ea958d1b5c67105f15f7e73dad1a0421b4f788ca/app/controllers/replies_controller.rb#L93L96, and for some reason, this did not feel bad. I need to rethink this.

**Comment by frenkel** [comment]
Hi,
Thanks for you report. This change was made intentionally because our customers were including other persons in their replies and were annoyed that we stopped including them. I do agree it is not transparent for client1 if he does not sign in to Brimir. If we would sent one notification instead of one notification _per user_, this problem would not occur, because the users would see all recipients in the to/cc fields.
Maybe we should only send the notification of the message client1 sent to the agents. If client1 wants to include client2 he can do so himself by adding him to the cc.

**Comment by fiedl** [comment]
I had a couple of days to think about that.

I do agree that the problem would not occur if the notification emails would have the `To` and `CC` fields filled in correctly. The notification system would still loop over all `notified_users`, but each `user` is set in `message.smtp_envelope_to` rather than in `message.to`.

Since the replying agent most likely does not want to reveal all recipients to the client — in particular, he does not want to expose other agent's real email addresses — the replying agent needs the ability to specify the `To` recipients, `CC` recipients and `BCC` recipients separately. (=> **`notifications.delivery_type`**)

(If we would want to keep things simpler, the system could automatically set all clients into the `To` field and all agents into the `BCC` field.)

Also, we need to track the delivery state in order to make sure that we only deliver notifications that have not been sent already by the sender's email server. (=> **`notifications.state`**)

```ruby
class Notification < ActiveRecord::Base
  belongs_to :notifiable, polymorphic: true
  belongs_to :user

  # Track the delivery type:
  #   (a) The recipient is specified in the `To` field.
  #   (b) The recipient is specified in the `CC` field.
  #   (c) The recipient is specified in the `BCC` field.
  #   (d) The recipient is notified by the ticket system only: If a client creates
  #       a new ticket via email, he does not include the agent direclty in any of
  #       the `To`, `CC`, or `BCC` fields. The agent is just notified by the ticket
  #       system.
  #
  enum delivery_type: {to: nil, cc: 1, bcc: 2, agent_notification: 3}
  scope :to, -> { where('delivery_type IS NULL') }
  
  # Track the state of the delivery.
  #   (a) The notification has been delivered somehow. This is the default
  #       since we can't determine the state of notifications before that code change.
  #   (b) The notification is due to deliver. Only messages that are `due` will be
  #       sent by brimir.
  #   (c) The notification has been delivered by brimir.
  #   (d) The notification has been delivered by the sender email server.
  #       If the client sends an email `To support@example.com, CC foo@example.com`
  #       then the recipient foo@example.com has already been notified by the sender
  #       server and does not need to be notified by brimir.
  #
  enum state: {delivered: nil, due: 1, delivered_by_brimir: 2, delivered_by_external_sender: 3}
  scope :delivered, -> { where('state IS NULL OR state = ? OR state = ?', 2, 3) }
end
```

I will try this out and then report the results.

**Comment by frenkel** [comment]
Hi,

Thanks for thinking about this. Wouldn't my second solution be a lot simpler? Sending only to agents when a client sends to Brimir. If the client wants to include other people he can always cc/bcc in the e-mail instead of letting Brimir handle that.

**Comment by fiedl** [comment]
I have experimented with the idea to let the user choose the `To`, `CC` and `BCC` flelds separately  in the web ui.

<img width="579" alt="bildschirmfoto_2016-02-09_um_20 47 16" src="https://cloud.githubusercontent.com/assets/1679688/13864035/b0fb0e14-ec9e-11e5-971e-d8ed732cc1ea.png">

<img width="576" alt="bildschirmfoto_2016-02-09_um_20 51 15" src="https://cloud.githubusercontent.com/assets/1679688/13864034/b08f9aee-ec9e-11e5-855b-ed944ba97bf7.png">

But, on second thought, this would just replace one problem with another: Where the agents would gain the ability to choose whom to notify transparently and whom to notify silently via `BCC`, the agents would also be responsible for never adding agents to the `To` or `CC` field in order to not expose their private email addresses.

For future reference, I've stored my experiments in https://github.com/fiedl/brimir/commit/78c4f96cdec39606e3017a3869f2c6a22e7a36b0.

But, on the second attempt, I followed your idea, @frenkel, to basically only send to agents whenever a clients replies via email. I'm going to create a pull request in a couple of minutes.
